crates/geoblock/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+                    Ok(())
+                } else {
+                    #[cfg(feature = "tracing")]
+                    {
+                        error!("Unable to extract geo data from IP address: {}", _e);
+                    }
+                    Err(GeoBlockError::UnableToExtractGeoData)
+                }
+            }
+        };
+
+        let is_blocked = self
+            .blocked_countries
+            .iter()
+            .any(|blocked_country| *blocked_country == country);
+
+        if is_blocked {
+            Err(GeoBlockError::BlockedCountry { country })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl<R, S, ReqBody, ResBody> Service<Request<ReqBody>> for GeoBlock<S, R>
+where
+    R: GeoIpResolver,
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body + Default,
+{
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+    type Response = Response<ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match self.extract_ip(&req) {
+            Ok(ip_addr) => match self.check_caller(ip_addr) {
+                Ok(_) => ResponseFuture::future(self.inner.call(req)),
+                Err(GeoBlockError::BlockedCountry { country: _country }) => {
+                    let mut res = Response::new(ResBody::default());
+                    *res.status_mut() = StatusCode::UNAUTHORIZED;
+                    ResponseFuture::invalid_ip(res)
+                }
+                Err(_e) => {
+                    let mut res = Response::new(ResBody::default());
+                    *res.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+                    ResponseFuture::invalid_ip(res)
+                }
+            },
+            Err(_e) => {
+                if self.missing_country == MissingCountry::Allow {
+                    #[cfg(feature = "tracing")]
+                    {
+                        error!("Unable to extract client IP address: {}", _e);
+                    }
+                    return ResponseFuture::future(self.inner.call(req));
+                }
+
+                let mut res = Response::new(ResBody::default());
+                *res.status_mut() = StatusCode::UNAUTHORIZED;
+                ResponseFuture::invalid_ip(res)
+            }
+        }
+    }
+}
+
+#[pin_project]
+/// Response future for [`GeoBlock`].
+pub struct ResponseFuture<F, B> {
+    #[pin]
+    inner: Kind<F, B>,
+}
+
+impl<F, B> ResponseFuture<F, B> {
+    fn future(future: F) -> Self {
+        Self {
+            inner: Kind::Future { future },
+        }
+    }
+
+    fn invalid_ip(res: Response<B>) -> Self {
+        Self {
+            inner: Kind::Error {
+                response: Some(res),
+            },
+        }
+    }
+}
+
+#[pin_project(project = KindProj)]
+enum Kind<F, B> {
```

Comment:
Actually, we don't need this. You may use an [`Either`](https://docs.rs/futures/latest/futures/future/enum.Either.html) here instead of a custom type. The left would be your future, while the right would be something like `std::future::ready()`. The less code the better.

Reply:
Wdym? It's a generic sum type of 2 futures, you can use it for anything you want. Anyway, will do